volatility/framework/plugins/windows/netscan.py

```diff
@@ -0,0 +1,158 @@
+             layer_name: str,
+             nt_symbol_table: str,
+             network_symbol_table: str) -> \
+        Iterable[interfaces.objects.ObjectInterface]:
+        """Scans for network objects using the poolscanner module and constraints.
+
+        Args:
+            context: The context to retrieve required elements (layers, symbol tables) from
+            layer_name: The name of the layer on which to operate
+            nt_symbol_table: The name of the table containing the kernel symbols
+            symbol_table: The name of the table containing the network object symbols (_TCP_LISTENER etc.)
+
+        Returns:
+            A list of network objects found by scanning the `layer_name` layer for network pool signatures
+        """
+
+        constraints = poolscanner.PoolScanner.builtin_constraints(network_symbol_table, [b'TcpL', b'TcpE', b'UdpA'])
+
+        for result in poolscanner.PoolScanner.generate_pool_scan(context, layer_name, nt_symbol_table, constraints):
+
+            _constraint, mem_object, _header = result
+            yield mem_object
+
+    def _generator(self):
+        netscan_symbol_table = self.create_netstat_symbol_table(self.context, self.config["nt_symbols"],
+                                                                self.config_path)
+
+        for netw_obj in self.scan(self.context, self.config['primary'], self.config['nt_symbols'],
+                                  netscan_symbol_table):
+
+            if isinstance(netw_obj, network._UDP_ENDPOINT):
+                if netw_obj.is_valid():
+                    vollog.debug(format_hints.Hex(netw_obj.vol.offset))
+                    if netw_obj.get_owner() != None:
+                        pid = int(netw_obj.get_owner().UniqueProcessId)
+                        proc_name = netw_obj.get_owner().ImageFileName.cast(
+                            "string", max_length = netw_obj.get_owner().ImageFileName.vol.count, errors = "replace")
+                    else:
+                        pid = 0
+                        proc_name = "-"
+                    # For UdpA, the state is always blank and the remote end is asterisks
+                    for ver, laddr, _ in netw_obj.dual_stack_sockets():
+                        yield (0, (format_hints.Hex(netw_obj.vol.offset), "UDP" + ver, "{}:{}".format(
+                            laddr, netw_obj.Port), "*:*", "", pid, proc_name, netw_obj.get_create_time()))
+
+            elif isinstance(netw_obj, network._TCP_ENDPOINT):
+                if netw_obj.is_valid():
+                    vollog.debug(format_hints.Hex(netw_obj.vol.offset))
+                    if netw_obj.get_address_family() == network.AF_INET:
+                        proto = "TCPv4"
+                    elif netw_obj.get_address_family() == network.AF_INET6:
+                        proto = "TCPv6"
+
+                    if netw_obj.get_owner() != None:
+                        pid = int(netw_obj.get_owner().UniqueProcessId)
```

Comment:
done and appears to work. However it might be useful to make it more stable by catching InvalidAddressExceptions when trying to dereference invalid pointers. On the other hand UniqueProcessId isn't a pointer and thus always contains *some kind* of value, which might be invalid but should not crash... thoughts?

Reply:
It's tricky, I think the compromise we've reached within the team is the addition of a has_valid_member method, which allows you to check whether the member is actually accessible or not.  It's not yet landed, but it's in the `has_valid_member` branch for testing.  We try to avoid catching exceptions excessively, particularly for code that will be used by other code rather than just the user, because they may need to know what's gone wrong and why.  Catching an exception says that you're taking responsibility for it in all situations.  In certain cases, that makes a lot of sense, but in vol2 we just indiscriminately said "that's fine, don't worry about it" which for a forensics tool seems a little too easy going...  5:P

I'm not sure we can vet all data in a memory dump, there is a degree where we simply report what's present so in the cases of UniqueProcessId, that's what it said, but if you start looking around it turns out that can't be right.  If we start deciding what we think is right, we'll end up building filters that malware may find ways to abuse (oh, volatility always ignores processes with a vastly higher pid, so we just do that and then we vanish).  As I say, it's a tricky one, but my own view is to show what you've been given and allow the code to figure out if something weird's gone on in the interval unless you're certain it's supposed to have happened...

Reply:
`has_valid_member` has now landed in master, so should be usable.  5:)